pkg/antctl/raw/traceflow/command.go

```diff
@@ -0,0 +1,250 @@
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package traceflow
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"k8s.io/klog"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v2"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+
+	"github.com/vmware-tanzu/antrea/pkg/antctl/runtime"
+	"github.com/vmware-tanzu/antrea/pkg/apis/ops/v1alpha1"
+	clientset "github.com/vmware-tanzu/antrea/pkg/client/clientset/versioned"
+	"github.com/vmware-tanzu/antrea/pkg/ops"
+)
+
+var (
+	Command     *cobra.Command
+	source      string
+	destination string
+	outputType  string
+	waiting     bool
+
+	protocol int32
+	srcPort  int32
+	dstPort  int32
+)
+
+// Response is the response of antctl traceflow.
+type Response struct {
+	Name        string                  `json:"name" yaml:"name"`      // Traceflow name
+	Phase       v1alpha1.TraceflowPhase `json:"phase,omitempty"`       // Traceflow phase
+	Source      string                  `json:"source,omitempty"`      // Traceflow source, e.g. "default/pod0"
+	Destination string                  `json:"destination,omitempty"` // Traceflow destination, e.g. "default/pod1"
+	NodeResults []v1alpha1.NodeResult   `json:"results,omitempty"`     // Traceflow node results
+}
+
+func init() {
+	Command = &cobra.Command{
+		Use:     "traceflow",
+		Short:   "Start a Traceflows",
+		Long:    "Start a Traceflows from one pod to another.",
+		Aliases: []string{"tf", "traceflows"},
+		Example: `  Start a Traceflow from busybox0 to busybox1, both pods are in namespace default
+  $antctl traceflow -S busybox0 -D busybox1
+  Start a Traceflow from busybox0 to destination IP, source is in namespace default
+  $antctl traceflow -S busybox0 -D 123.123.123.123
+  Start a Traceflow from busybox0 in namespace ns0 to busybox1 in namespace ns1, output type is json
+  $antctl traceflow -S ns0/busybox0 -D ns1/busybox1 -o json
+  Start a Traceflow from busybox0 to busybox1, with TCP header and 80 as destination port
+  $antctl traceflow -S busybox0 -D busybox1 -p 6 --dstport 80
```

Comment:
I think we should allow users to use tcp/udp/icmp/sctp directly.

Again would you consider OVS flow expression? Check "Flow Syntax" in http://www.openvswitch.org/support/dist-docs/ovs-ofctl.8.txt? I think it is straightforward to use, with the extra benefits of being consistent with antctl trace-packet, and all OVS commnds.
We just need to support a subset of the flow fields we support, like tcp, udp, icmp, tcp_dst, udp_src, nw_proto, sctp, etc.

Like: antctl traceflow -S busybox0 -D busybox1 -f tcp,tcp_dst=80

Reply:
Thank you Jianjun! Actually I don't have strong preference. ovs-tracing command applies "ovs-appctl". Since flow is its argument, it makes sense to use flow expression in antctl. While traceflow is using CRD to pass config to controller and agent. As for implementation, they are different.

However, you mentioned the consistency between antctl commands. Then I think yours is better.

Reply:
Yes, consistent design is one thing. I also feel OVS flow syntax is convenient to use.